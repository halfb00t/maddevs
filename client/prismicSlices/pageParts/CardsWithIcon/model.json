{
  "id": "cards_with_icon",
  "type": "SharedSlice",
  "name": "CardsWithIcon",
  "description": "CardsWithIcon",
  "variations": [
    {
      "id": "default-slice",
      "name": "Default slice",
      "docURL": "...",
      "version": "sktwi1xtmkfgx8626",
      "description": "CardsWithIcon",
      "primary": {},
      "items": {
        "iconImage": {
          "config": {
            "label": "Icon",
            "constraint": {},
            "thumbnails": []
          },
          "type": "Image"
        },
        "title": {
          "config": {
            "label": "Title",
            "placeholder": "",
            "allowTargetBlank": true,
            "single": "heading1,heading2,heading3,heading4,heading5,heading6,strong"
          },
          "type": "StructuredText"
        },
        "description": {
          "config": {
            "label": "Description",
            "placeholder": ""
          },
          "type": "Text"
        }
      }
    }
  ]
}