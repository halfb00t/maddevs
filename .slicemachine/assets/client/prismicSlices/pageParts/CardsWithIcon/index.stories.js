import MyComponent from '../../../../../../client/prismicSlices/pageParts/CardsWithIcon';
import SliceZone from 'vue-slicezone'

export default {
  title: 'client/prismicSlices/pageParts/CardsWithIcon'
}


export const _DefaultSlice = () => ({
  components: {
    MyComponent,
    SliceZone
  },
  methods: {
    resolve() {
      return MyComponent
    }
  },
  data() {
    return {
      mock: {"variation":"default-slice","name":"Default slice","slice_type":"cards_with_icon","items":[{"iconImage":{"dimensions":{"width":900,"height":500},"alt":"Placeholder image","copyright":null,"url":"https://images.unsplash.com/photo-1544731612-de7f96afe55f?w=900&h=500&fit=crop"},"title":[{"type":"heading1","text":"Integrate proactive metrics","spans":[]}],"description":"envisioneer interactive synergies"},{"iconImage":{"dimensions":{"width":900,"height":500},"alt":"Placeholder image","copyright":null,"url":"https://images.unsplash.com/photo-1576662712957-9c79ae1280f8?w=900&h=500&fit=crop"},"title":[{"type":"heading1","text":"Engage B2B platforms","spans":[]}],"description":"cultivate real-time e-commerce"}],"primary":{},"id":"_DefaultSlice"}
    }
  },
  template: '<SliceZone :slices="[mock]" :resolver="resolve" />'
})
_DefaultSlice.storyName = 'Default slice'
