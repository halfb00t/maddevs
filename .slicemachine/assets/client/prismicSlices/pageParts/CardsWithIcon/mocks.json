[
  {
    "variation": "default-slice",
    "name": "Default slice",
    "slice_type": "cards_with_icon",
    "items": [
      {
        "iconImage": {
          "dimensions": {
            "width": 900,
            "height": 500
          },
          "alt": "Placeholder image",
          "copyright": null,
          "url": "https://images.unsplash.com/photo-1515378960530-7c0da6231fb1?w=900&h=500&fit=crop"
        },
        "title": [
          {
            "type": "heading1",
            "text": "Deliver strategic deliverables",
            "spans": []
          }
        ],
        "description": "grow front-end schemas"
      },
      {
        "iconImage": {
          "dimensions": {
            "width": 900,
            "height": 500
          },
          "alt": "Placeholder image",
          "copyright": null,
          "url": "https://images.unsplash.com/photo-1515378960530-7c0da6231fb1?w=900&h=500&fit=crop"
        },
        "title": [
          {
            "type": "heading1",
            "text": "Seize best-of-breed e-commerce",
            "spans": []
          }
        ],
        "description": "brand sexy markets"
      },
      {
        "iconImage": {
          "dimensions": {
            "width": 900,
            "height": 500
          },
          "alt": "Placeholder image",
          "copyright": null,
          "url": "https://images.unsplash.com/photo-1544731612-de7f96afe55f?w=900&h=500&fit=crop"
        },
        "title": [
          {
            "type": "heading1",
            "text": "Extend viral web-readiness",
            "spans": []
          }
        ],
        "description": "integrate dynamic blockchains"
      },
      {
        "iconImage": {
          "dimensions": {
            "width": 900,
            "height": 500
          },
          "alt": "Placeholder image",
          "copyright": null,
          "url": "https://images.unsplash.com/photo-1493119508027-2b584f234d6c?w=900&h=500&fit=crop"
        },
        "title": [
          {
            "type": "heading1",
            "text": "Transform interactive niches",
            "spans": []
          }
        ],
        "description": "extend ubiquitous networks"
      },
      {
        "iconImage": {
          "dimensions": {
            "width": 900,
            "height": 500
          },
          "alt": "Placeholder image",
          "copyright": null,
          "url": "https://images.unsplash.com/photo-1587905069134-008460d7a636?w=900&h=500&fit=crop"
        },
        "title": [
          {
            "type": "heading1",
            "text": "Aggregate next-generation vortals",
            "spans": []
          }
        ],
        "description": "innovate customized portals"
      }
    ],
    "primary": {
      "animation": "zoom-in",
      "containerSize": "1240"
    }
  }
]