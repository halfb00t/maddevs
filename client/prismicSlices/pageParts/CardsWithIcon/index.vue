<template>
  <section class="section cards-with-icons">
    <div
      class="cards-with-icons__wrapper container"
      :data-aos="animation || 'fade-up'"
      :style="`max-width: ${containerSize || '1240'}px`"
    >
      <div
        v-for="(item, i) of items"
        :key="i"
        class="cards-with-icons__item"
      >
        <img
          v-if="item.iconImage.url"
          v-lazy-load
          :data-src="item.iconImage.url"
          width="54"
          height="54"
          :alt="item.iconImage.alt || 'Icon image'"
          class="cards-with-icons__item-icon"
        >
        <PrismicRichText
          v-if="$prismic.asText(item.title)"
          :field="item.title"
          data-testid="cards-with-icons__item-title"
          class="cards-with-icons__item-title"
          :html-serializer="serializer"
        />
        <p class="cards-with-icons__item-description">
          {{ item.description }}
        </p>
      </div>
    </div>
  </section>
</template>

<script>
import convertTagsToText from '@/helpers/convertTagsToText'

export default {
  name: 'CardsWithIcon',
  props: {
    slice: {
      type: Object,
      required: true,
      default() {
        return {}
      },
    },
  },

  data() {
    return {
      items: this.slice?.items,
      animation: this.slice?.primary?.animation,
      containerSize: this.slice?.primary?.containerSize,
    }
  },

  computed: {
    serializer() {
      const types = ['cards_with_icon']
      if (types.includes(this.slice?.slice_type)) return this.htmlSerializer
      return null
    },
  },

  methods: {
    htmlSerializer(type, element, content, children) {
      const { Elements } = this.$prismic.dom.RichText
      let text = children.join('')
      text = text.replace(/`(.*?)`/g, (_, inlineCode) => {
        // the second parameter of function excludes tags
        const formattedCode = convertTagsToText(inlineCode, ['strong', 'em', 'a'])
        return `<code class="inline-code">${formattedCode}</code>`
      })

      switch (type) {
        case Elements.heading1: return `<h1>${text}</h1>`
        case Elements.heading2: return `<h2>${text}</h2>`
        case Elements.heading3: return `<h3>${text}</h3>`
        case Elements.heading4: return `<h4>${text}</h4>`
        case Elements.heading5: return `<h5>${text}</h5>`
        case Elements.heading6: return `<h6>${text}</h6>`
        case Elements.strong: return `<strong>${text}</strong>`
        default: return null
      }
    },
  },
}
</script>

<style lang="scss" scoped>
.cards-with-icons {
  &__wrapper {
    @include grid(repeat(3, 1fr), auto, 30px, 30px);

    @media screen and (max-width: 768px) {
      @include grid(repeat(2, 1fr), auto, 30px, 30px);
    }

    @media screen and (max-width: 540px) {
      @include grid(repeat(1, 1fr), auto, 0, 30px);
    }
  }

  &__item {
    background-color: $bgcolor--black-pale;
    padding: 56px 54px;
    border-radius: 4px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: flex-start;
    @media screen and (max-width: 1024px) {
      padding: 24px;
    }
  }

  &__item-title {
    margin-bottom: 21px;
    color: $text-color--white-primary;
    line-height: 32.5px;
    letter-spacing: -0.013em;
    ::v-deep h1,
    ::v-deep h2,
    ::v-deep h3,
    ::v-deep h4,
    ::v-deep h5,
    ::v-deep h6 {
      @include font('Inter', 28px, 600);
      @media screen and (max-width: 1024px) {
        @include font('Inter', 21px, 600);
        line-height: 24px;
        letter-spacing: -1px;
      }
    }
  }

  &__item-description {
    @include font('Inter', 16px, 400);
    line-height: 24px;
    letter-spacing: -0.4px;
    color: $text-color--grey-20-percent;
  }

  &__item-icon {
    width: 54px;
    height: 54px;
    margin-bottom: 36px;
  }
}
</style>
