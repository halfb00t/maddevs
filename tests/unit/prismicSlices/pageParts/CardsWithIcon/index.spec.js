import { render, screen } from '@testing-library/vue'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import CardsWithIcon from '@/prismicSlices/pageParts/CardsWithIcon'
// eslint-disable-next-line import/order
import lazyLoad from 'nuxt-lazy-load/lib/module'

const props = {
  slice: {
    primary: {
      animation: 'fade-up',
      containerSize: '1240',
    },
    items: [{
      title: [{
        text: 'Test title',
        type: 'heading4',
      }],
      iconImage: {
        url: 'img.jpg',
        alt: 'Test image',
      },
      description: 'Test description',
    }],
    slice_type: 'cards_with_icon',
  },
}

const mocks = {
  $prismic: {
    asText: text => text[0].text,
  },
}

const localVue = createLocalVue()
localVue.directive('lazy-load', lazyLoad)
jest.mock('nuxt-lazy-load/lib/module')

const stubs = ['PrismicRichText']

describe('CardsWithIcon component', () => {
  let wrapper = null

  beforeEach(() => {
    wrapper = shallowMount(CardsWithIcon, {
      propsData: props,
      mocks,
      localVue,
      stubs,
    })
  })

  afterEach(() => {
    wrapper = null
  })

  it('should correctly render component', () => {
    const container = render(CardsWithIcon, {
      propsData: props,
      localVue,
      mocks,
      stubs,
    })

    expect(screen.queryByTestId('cards-with-icons__item-title')).not.toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should return empty object from slice props', () => {
    expect(wrapper.vm.$options.props.slice.default.call()).toEqual({})
  })

  it('should return null from serializer computed method', () => {
    wrapper.vm.slice.slice_type = 'unknown'
    expect(wrapper.vm.serializer).toBeNull()
  })
})

describe('htmlSerializer', () => {
  let wrapper = null

  beforeEach(() => {
    wrapper = shallowMount(CardsWithIcon, {
      stubs,
      propsData: props,
      localVue,
      mocks: {
        $prismic: {
          asText: text => text[0].text,
          dom: {
            RichText: {
              Elements: {
                heading1: 'heading1',
                heading2: 'heading2',
                heading3: 'heading3',
                heading4: 'heading4',
                heading5: 'heading5',
                heading6: 'heading6',
                strong: 'strong',
              },
            },
          },
        },
      },
    })
  })

  afterEach(() => {
    wrapper = null
  })

  it('if type not heading will return null', () => {
    const result = wrapper.vm.htmlSerializer('div', null, null, ['CardsWithIcon title'])
    expect(result).toBeNull()
  })

  it('if type heading1 will return html with h1 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading1', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h1>CardsWithIcon title</h1>')
  })

  it('if type heading2 will return html with h2 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading2', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h2>CardsWithIcon title</h2>')
  })

  it('if type heading3 will return html with h3 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading3', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h3>CardsWithIcon title</h3>')
  })

  it('if type heading4 will return html with h4 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading4', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h4>CardsWithIcon title</h4>')
  })

  it('if type heading5 will return html with h5 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading5', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h5>CardsWithIcon title</h5>')
  })

  it('if type heading6 will return html with h6 tag', () => {
    const result = wrapper.vm.htmlSerializer('heading6', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<h6>CardsWithIcon title</h6>')
  })

  it('if type heading6 will return html with pre code text', () => {
    const result = wrapper.vm.htmlSerializer('heading6', null, null, ['`<em>ololo</em>` CardsWithIcon title'])
    expect(result).toBe('<h6><code class="inline-code"><em>ololo</em></code> CardsWithIcon title</h6>')
  })

  it('if type strong will return <strong>CardsWithIcon title</strong>', () => {
    const result = wrapper.vm.htmlSerializer('strong', null, null, ['CardsWithIcon title'])
    expect(result).toBe('<strong>CardsWithIcon title</strong>')
  })
})
