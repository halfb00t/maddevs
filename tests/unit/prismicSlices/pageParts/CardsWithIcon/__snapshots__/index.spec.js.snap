// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsWithIcon component should correctly render component 1`] = `
Object {
  "baseElement": <body>
    <div>
      <section
        class="section cards-with-icons"
      >
        <div
          class="cards-with-icons__wrapper container"
          data-aos="fade-up"
          style="max-width: 1240px;"
        >
          <div
            class="cards-with-icons__item"
          >
            <img
              alt="Test image"
              class="cards-with-icons__item-icon"
              data-src="img.jpg"
              height="54"
              width="54"
            />
             
            <prismicrichtext-stub
              class="cards-with-icons__item-title"
              data-testid="cards-with-icons__item-title"
              field="[object Object]"
              html-serializer="function () { [native code] }"
            />
             
            <p
              class="cards-with-icons__item-description"
            >
              
        Test description
      
            </p>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="section cards-with-icons"
    >
      <div
        class="cards-with-icons__wrapper container"
        data-aos="fade-up"
        style="max-width: 1240px;"
      >
        <div
          class="cards-with-icons__item"
        >
          <img
            alt="Test image"
            class="cards-with-icons__item-icon"
            data-src="img.jpg"
            height="54"
            width="54"
          />
           
          <prismicrichtext-stub
            class="cards-with-icons__item-title"
            data-testid="cards-with-icons__item-title"
            field="[object Object]"
            html-serializer="function () { [native code] }"
          />
           
          <p
            class="cards-with-icons__item-description"
          >
            
        Test description
      
          </p>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "isUnmounted": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "unmount": [Function],
  "updateProps": [Function],
}
`;
